### ISSUE TITLE
Role mapping not updating kubernetes_users

### ISSUE THREAD

**Issue by kgtw** [gravitational/teleport#4990]
### Description

**What happened**:

If we have set two values for `kubernetes_users` in both the root cluster role and a leaf cluster role, and then we remove one of the values in both clusters, kubectl commands fail to work as normal. Teleport is expecting you to provide who to impersonate via the `--as` argument of kubectl.

Logging in & out doesn't fix the issue either.

```
$ kubectl get nodes -v=9
I1125 21:45:12.280698    1344 loader.go:375] Config loaded from file:  /Users/demo/.kube/config
I1125 21:45:12.287810    1344 round_trippers.go:423] curl -k -v -XGET  -H "Accept: application/json;as=Table;v=v1beta1;g=meta.k8s.io, application/json" -H "User-Agent: kubectl/v1.17.9 (darwin/amd64) kubernetes/4fb7ed1" 'https://teleport.demo:3026/api/v1/nodes?limit=500'
I1125 21:45:12.460677    1344 round_trippers.go:443] GET https://teleport.demo:3026/api/v1/nodes?limit=500 403 Forbidden in 172 milliseconds
I1125 21:45:12.460726    1344 round_trippers.go:449] Response Headers:
I1125 21:45:12.460738    1344 round_trippers.go:452]     Cache-Control: no-cache, no-store, must-revalidate
I1125 21:45:12.460748    1344 round_trippers.go:452]     Cache-Control: no-cache, no-store, must-revalidate
I1125 21:45:12.460757    1344 round_trippers.go:452]     Content-Length: 130
I1125 21:45:12.460765    1344 round_trippers.go:452]     Content-Type: application/json
I1125 21:45:12.460773    1344 round_trippers.go:452]     Expires: 0
I1125 21:45:12.460780    1344 round_trippers.go:452]     Expires: 0
I1125 21:45:12.460788    1344 round_trippers.go:452]     Pragma: no-cache
I1125 21:45:12.460798    1344 round_trippers.go:452]     Pragma: no-cache
I1125 21:45:12.460820    1344 round_trippers.go:452]     Date: Wed, 25 Nov 2020 21:45:12 GMT
I1125 21:45:12.460939    1344 request.go:1017] Response Body: {"message":"please select a user to impersonate, refusing to select a user due to several kuberenetes_users set up for this user"}
I1125 21:45:12.461162    1344 request.go:1220] body was not decodable (unable to check for Status): Object 'Kind' is missing in '{"message":"please select a user to impersonate, refusing to select a user due to several kuberenetes_users set up for this user"}'
I1125 21:45:12.461364    1344 helpers.go:203] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "unknown (get nodes)",
  "reason": "Forbidden",
  "details": {
    "kind": "nodes",
    "causes": [
      {
        "reason": "UnexpectedServerResponse",
        "message": "unknown"
      }
    ]
  },
  "code": 403
}]
F1125 21:45:12.461396    1344 helpers.go:114] Error from server (Forbidden): unknown (get nodes)
```

before root cluster
```
  allow:
    kubernetes_users:
    - '{{external.username}}'
    - 'user-to-be-removed@email.com'
    logins:
    - '{{email.local(external.username)}}'
```

before leaf cluster
```
  allow:
    kubernetes_users:
    - '{{internal.username}}'
    - 'user-to-be-removed@email.com'
    logins:
    - '{{internal.logins}}'
```

after root cluster
```
  allow:
    kubernetes_users:
    - '{{external.username}}'
    logins:
    - '{{email.local(external.username)}}'
```

after leaf cluster
```
  allow:
    kubernetes_users:
    - '{{internal.username}}'
    logins:
    - '{{internal.logins}}'
```

**What you expected to happen**:

Roles with a single `kubernetes_user` specified to have normal access without needing to use `--as` anymore.

**How to reproduce it (as minimally and precisely as possible)**:

Create a role in root & leaf clusters with two values for `kubernetes_users`. With this you will have kubectl access by using `--as` argument to specify which user to use.

Remove one of the users from `kubernetes_users` in both root & leaf cluster, kubectl is still asking you who to impersonate. Since there is only a single user allowed this should be the default used.

### Environment
- Teleport version (use `teleport version`): 4.4.0 Enterprise
- Tsh version (use `tsh version`): 4.4.0 Enterprise
- OS (e.g. from `/etc/os-release`):

- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): AWS

**Browser environment**

- Browser Version (for UI-related issues):
- Install tools:
- Others:

**Relevant Debug Logs If Applicable**

- `tsh --debug`
- `teleport --debug`

From leaf cluster
```
ERRO [PROXY:KUB] Failed to set up forwarding headers: please select a user to impersonate, refusing to select a user due to several kuberenetes_users set up for this user. proxy/forwarder.go:908
```

**Comment by kgtw** [comment]
After some time (maybe 1hr?) it starts working as expected. I'm guessing some local cache in the leaf cluster isn't being flushed when roles are updated?

**Comment by awly** [comment]
@kgtw `{{internal.username}}` in your leaf role is not valid, so you end up with empty `kubernetes_users`
Can you try `{{internal.logins}}` or `{{internal.kubernetes_users}}` in the leaf role?

**Comment by awly** [comment]
Ahh, I think I see another problem.
When we set impersonation headers, we look at the number of `kubernetes_users` in the actual user role, not in the cert. I guess when you updated the role, proxy somehow used the stale version.

**Comment by kgtw** [comment]
Oh sorry, a case of bad copy & paste. We are using `{{internal.logins}}` and `{{internal.kubernetes_users}}` on the leaf role.

Yes, I think it is checking the role not the cert as I logged out and in again and it was a persistent issue. With my new cert I could only see the one kubernetes user as expected.

If I were to hazard a guess it feels like it is checking the role, and somehow the role check is cached somehow as the updated role was not being reflected in what was actually happening, even with tctl and dynamodb showing that the role is in the correct state.

**Comment by awly** [comment]
I can not reproduce this with 5.1.

My repro steps:
- set up 2 clusters, root and leaf
- create this role in both clusters:
```
kind: role
version: v3
metadata:
  name: admin
spec:
  allow:
    kubernetes_groups:
    - '{{internal.kubernetes_groups}}'
    kubernetes_users:
    - 'test-user1'
    - 'test-user2'
    logins:
    - '{{internal.logins}}'
  deny:
```
- create a trusted cluster mapping `admin` role from root to same role on leaf
- log into root, switch to leaf
- confirm that `kubectl --as` is required to choose one of the usernames
- remove `test-user2` from roles in both clusters
- `tsh logout; tsh login ...`
- confirm that `kubectl` against root cluster no longer asks for `--as`
- switch to leaf and confirm that `kubectl` against leaf cluster no longer asks for `--as`

@kgtw LMK if my repro was incorrect.
Otherwise, I'm closing this bug, assuming that it was fixed during the big k8s refactor of 5.0 release.

**Comment by kgtw** [comment]
Thanks for checking this @awly. I'll try to reproduce when we upgrade to 5.x
